Userland/Libraries/LibThreading/BackgroundAction.h

```diff
@@ -100,7 +100,7 @@ class BackgroundAction final : public Core::Object
                     error = result.release_error();
 
                 m_promise->cancel(Error::from_errno(ECANCELED));
-                if (m_on_error) {
+                if (!m_canceled && m_on_error) {
```

Comment:
No callback being called in case of cancellation looks off to me, and I don't understand how it resolves an UAF (more explanations in the commit message would be welcome).

Anyway, `FileSystemModel` currently rely on it, so it needs some changes:
```cpp
s_thumbnail_cache.with_locked([path, error = move(error)](auto& cache) {
    if (error != Error::from_errno(ECANCELED)) {
        cache.thumbnail_cache.set(path, nullptr);
        dbgln("Failed to load thumbnail for {}: {}", path, error);
    }
    cache.loading_thumbnails.remove(path);
});
```

Reply:
The UAF is of the event loop, which may be gone by the time the background action ends.

Reply:
The only solution here is to call the error callback synchronously as before. This should be fine, since we had that behavior before and anyone doing a lot of work in the error callback should then take care to not destroy the event loop.